File: LinkedLists/C++/Eliminaciones.cpp
#include<bits/stdc++.h>
using namespace std;

//Creamos la clase nodo en donde damos la estructura de lo que contiene un nodo con dos constructores disponibles
class Node{
    public:
        int data; //Dato que contiene el nodo
        Node *next; //Puntero al siguiente nodo
        Node(){this->data = 0 ; this->next = NULL;}
        Node(int data){this->data = data ; this->next = NULL;}
};

//Creamos la clase lista la cual contendra la cabeza de la lista y un atributo adicional para llevar un control del tamano de la lista
class LinkedList{
    public:
        int size; //Atributo que se ira actualizando con el tamano de la lista 
        Node *head; //Creamos un puntero hacia el nodo cabeza de nuestra lista
        LinkedList(){this->size = 0 ; this->head = new Node();} //Constructor
};

//Funcion para eliminar el primer nodo de una lista
//Complejidad Temporal: O(1)
//Complejidad Espacial: O(1)
//Esta funcion recibira el puntero a la lista, y la referencia del nodo cabeza
void pop(LinkedList *list, Node *&head){
    //Verificamos que la lista no este vacia
    if(list->size > 0){
        //Guardaremos la referencia del primer elemento para no perderla luego de hacer las otras operaciones
        Node *temp = head->next;
        //El primer elemento de la lista ahora debe de ser el siguiente del actual primero
        //Es decir, el siguiente de cabeza debe de apuntar al siguiente del primer elemento
        head->next = head->next->next;
        //Ahora, como guardamos en el nodo temp, la referencia al nodo que fue el primero de la lista, lo debemos desvincular de ella, por lo que pondremos a apuntar como su siguiente a nulo
        temp->next = NULL;
        //Decrementamos la longitud de la lista
        list->size--;
        //Eliminamos esta referencia de memoria ya que no la utilizaremos mas 
        delete temp;
    }
}

//Funcion para eliminar ultimo nodo de una lista enlazada
//Complejidad Temporal: O(n)
//Complejidad Espacial: O(1)
//Esta funcion recibira el puntero a la lista, y la referencia del nodo cabeza
void popLast(LinkedList *list, Node *&head){
    //Verificamos que la lista no este vacia
    if(list->size > 0){
        //Creamos un nodo que nos servira como nodo actual, y este lo inicializaremos en el primer nodo, es decir, el siguiente de la cabeza
        Node *curr = head->next;
        //Crearemos otro nodo el cual servira para siempre tener una referencia al nodo previo de curr
        Node *prev = head;
        //Iteramos hasta que el siguiente de curr sea nulo. Esto quiere decir que curr se encuentra en el ultimo nodo de la lista y alli es donde debemos de parar para realizar las respectivas operaciones de eliminado de ese nodo
        while(curr->next != NULL){
            //Corremos ambos nodos a sus siguientes , empezando actualizando el prev
            prev = curr;
            curr = curr->next;
        }
        //Una vez finalizado el ciclo, sabemos que curr esta actualmente en el ultimo, y prev es el nodo antes de este
        //Antes de eliminar este ultimo nodo, debemos de asegurarnos que la lista siga su flujo normal, por lo que el previo debe de apuntar al siguiente de curr
        prev->next = curr->next;
        //Desvincularemos de la lista a curr haciendolo apuntar a nulo
        curr->next = NULL;
        //Decrementamos la longitud de la lista
        list->size--;
        //Eliminamos de memoria curr, el cual correspondia al ultimo nodo de la lista
        delete curr;
    }
}

//Funcion para eliminar ultimo nodo de la lista recursivamente
//Complejidad Temporal : O(n)
//Complejidad Espacial : O(n)
//Esta funcion recibira el puntero a la lista, la referencia al nodo cabeza, y dos nodos auxiliares prev y curr, en donde llegara siendo el primer nodo de la lista y prev, la cabeza.
void _popLast(LinkedList *list, Node *&head, Node *&prev, Node *&curr){
    //Caso Base:
    //Si el nodo curr, el cual es el que se va a mover en el algoritmo junto con prev, si el siguiente de curr es nulo, quiere decir que ha llegado al ultimo nodo de la lista, por lo que ahi toca parar y hacer la operacion de eliminar
    if(curr->next == NULL){
        //En este punto, antes de eliminar el ultimo nodo , es decir, curr, debemos de asegurarnos de que la lista siga conectada y no pierda su flujo
        //Para esto haremos que prev, apunte al siguiente de curr, lo que hara que la lista siga normal, y ya no se tenga acceso a curr pr ningun lado de la lista
        prev->next = curr->next;
        //Desvinculamos a curr de la lista haciendolo apuntar a nulo
        curr->next = NULL;
        //Decrementamos el valor de la longitud de la lista
        list->size--;
        //Eliminamos el ultimo nodo de memoria, es decir, a curr
        delete curr;
        return;
    }else{
        //Si el caso base no se cumple, significa que no hemos llegado al final de la lista, por lo que toca avanzar los dos nodos auxiliares a sus siguientes
        prev = curr;
        curr = curr->next;
        //Con estos datos actualizados, llamamos recursivamente la funcion con estos nuevos valores
        _popLast(list, head, prev, curr);
    }
}

//Funcion para eliminar un nodo en un posicion determinada
//Complejidad Temporal: O(n)
//Complejidad Espacial: O(1)
//Esta funcion recibira el puntero a la lista, la referencia del nodo cabeza, y la posicion
void erase(LinkedList *list, Node *&head, int pos){
    //Verificamos que la lista no este vacia y que la posicion no sea ni mayor ni menor que la longitud de la lista
    if(head != NULL && pos <= list->size && pos > 0 ){
        //Crearemos un contador empezado desde 1 para asi llevar la cuenta de en que nodo nos encontramos a medida que iteramos
        int cont = 1;
        //Crearemos un nodo curr el cual empezara desde el primer nodo de la lista, ya que este siempre debe de corresponder con la posicion que indica el contador. Ademas, este nos ayudara a tener la referencia al nodo que se encuentra en la posicion recibida.
        Node *curr = head->next;
        //Por otro lado, crearemos un nodo prev el cual nos servira para siempre tener referencia al nodo previo de curr, para asi poder hacer bien el proceso de eliminado cuando curr se encuentre en el nodo objetivo.
        Node *prev = head;
        //Iteraremos desde 1, hasta la el valor de la posicion recibida
        for (int i = 1; i <= pos; i++)
        {
            //Preguntamos si contador es igual a la posicion para asi saber si ya estamos en el nodo objetivo
            if(cont == pos){
                //Si se entra aca, significa que ya curr esta sobre el nodo objetivo y prev anterior a el
                //Antes de eliminar el nodo que nos interesa, debemos de asegurarnos que la lista siga su flujo.
                //Para esto, haremos que prev apunte al siguiente de curr
                prev->next = curr->next;
                //Con lo anterior, la lista sigue estando bien conectada, y ya on se tiene acceso por nigun lado de la lista a curr
                //Ahora pasaremos a desvincular a curr de la lista haciendolo apuntar a nulo
                curr->next = NULL;
                //Decrementamos la longitud de la lista
                list->size--;
                //Eliminamos de memoria a curr, es decir, al nodo objetivo
                delete curr;
            }else{
                //Si no entro a la anterior condicion, significa que aun curr no ha llegado al objetivo, por lo que debemos de aumentar el contador, y por consiguiente, los dos nodos auxiliares pasarlos a sus siguientes
                cont++;
                prev = curr;
                curr = curr->next;
            }
        }
        
    }
}

//-------------------------------------------

File: LinkedLists/Python/Eliminaciones.py
#Creamos la clase nodo en donde damos la estructura de lo que contiene un nodo con dos constructores disponibles
class Node:
    def __init__(self): #Constructor vacio
        self.data = None #Dato que almacenara el nodo
        self.next = None #Nodo siguiente
    
    def __init__(self,data): #Constructor que recibe el valor del nodo
        self.data = data
        self.next = None

#Creamos la clase lista la cual contendra la cabeza de la lista y un atributo para llevar el control del tamano de la lista
class LinkedList:
    def __init__(self):
        self.head = Node(None)
        self.size = 0

#Funcion para eliminar primer nodo de la lista
#Complejidad Temporal: O(1)
#Complejidad Espacial: O(1)
#Esta funcion recibira la lista y su cabeza
def pop(List, head):
    #Verificamos que la lista no este vacia
    if List.size > 0:
        #Creamos un nodo temporal que guarde la referencia al primer nodo de la lista, para luego poder acceder a el y realizar otras operaciones
        temp = head.next
        #Como necesitamos eliminar el primer nodo de la lista, antes de hacerlo, debemos de hacer que el la cabeza, la cual siempre apunta al primer elemento, ahora debe de apuntar , al siguiente del primer nodo actual.
        head.next = head.next.next
        #Con la cabeza ya apuntando al nuevo primer nodo, el antiguo primer nodo, lo desvincularemos de la lista, es decir, lo pondremos a apuntar a nulo
        temp.next = None
        #Decrementaremos el valor de la longitud de la lista
        List.size -= 1
        #Eliminamos la referencia al antiguo primer nodo
        del temp

#Funcion para eliminar ultimo nodo de una lista 
#Complejidad: O(n)
#Complejidad: O(1)
#Esta funcion recibira la lista, y su cabeza
def popLast(List, head):
    #Verificamos que la lista no este vacia
    if List.size > 0:
        #Creamos un nodo auxliar que nos servira como nodo actual a la hora de recorrer la lista. Este lo inicializaremos en el primer nodo de la lista
        curr = head.next
        #Ahora necesitamos tener un nodo que nos sirva para tener siempre una referencia al nodo previo del nodo curr. En este caso, empezaremos inicializando a prev en la cabeza
        prev = head
        #Ahora, iteraremos hasta que el siguiente de curr sea nulo, es decir, hasta que curr se encuentre en el ultimo nodo de la lista y a su vez tengamos referencia de su previo , con el nodo prev
        while curr.next is not None:
            #Ahora tenemos que empezar a pasar nuestros nodos auxiliares a sus siguiente, empezando actualizando a prev
            prev = curr
            curr = curr.next
        #Acabado el ciclo, sabemos que ya curr se encuentra en el ultimo nodo, y prev es su previo
        #Antes de pasar a eliminar a curr, es decir, el ultimo nodo de la lista, debemos de asegurarnos de que la lista siga su flujo normal. Para haremos que prev apunte ahora al siguiente de curr. Esto desvinculara a prev de curr, y asi la lista tendra un nuevo flujo en donde no accede a curr
        prev.next = curr.next
        #Ahora debemos de desvincular a curr de la lista, haciendolo apuntar a nulo
        curr.next = None
        #Decrementamos el valor de la longitud de la lista
        List.size -= 1
        #Eliminamos a curr de memoria, es decir, el ultimo nodo
        del curr

#Funcion para eliminar ultimo nodo de una lista recursivamente
#Comlejidad Temporal: O(n)
#Complejidad Espacial :O(n)
#Esta funcion recibira la lista , la cabeza, y dos nodos auxiliares: prev y curr. El primero llegare inicializado con la referencia de la cabeza de la lista y el segundo llegare con la referencia del primer nodo de a lista
def _popLast(List, head, prev, curr):
    #Caso base:
    #Debemos de parar cuando estemos ubicados en el utlimo ya que alli es donde se tendra que realizar la operacion de eliminar.
    #Para esto iremos moviendo a curr durante toda la lista hasta que llegue al ultimo elemento, es decir, hasta que su siguiente sea nulo.
    #De igual manera, moveremos a prev simultaneamente ya que este nos servira para tener una referencia al nodo pervio de curr
    if curr.next is None:
        #En este punto sabemos que curr esta en el ultimo nodo y prev esta antes de el.
        #Antes de pasar a eliminar el ultimo nodo, es decir a curr, debemos de asegurarnos de la que lista siga teniendo un flujo normal, y para esto si eliminamos a curr, dejaremos a prev sin ningun siguiente establecid.
        #Por ello, debemos de decir que e siguiente de prev sera el siguiente de curr. De esta manera ya no tendremos acceso a curr por niguna nodo de la lista
        prev.next = curr.next
        #Desvincularemos a curr de la lista haciendolo apuntar a nulo
        curr.next = None
        #Decrementamos la longitud de la lista
        List.size -=1
        #Eliminamos de memoria curr, es decir, el ultimo nodo
        del curr
        return
    else:
        #Si no se entro al caso base, significa que curr aun ha llegado al ultimo nodo. Por lo que debemos de actualizarlo y pasarlo a su siguiente.
        #Sin embargo, siempre necesitamos la referencia al nodo previo de curr, por lo tanto, tambien debemos de actualizar a prev
        prev = curr
        curr = curr.next
        #Ahora que los datos estan actualizados, podemos llamar la funcion recursivamente pasandole estos nuevos datos
        _popLast(List, head, prev, curr)

#Funcion para eliminar nodo en una posicion recibida de una lista
#Complejidad Temporal: O(n)
#Complejidad Espacial: O(1)
#Esta funcion recibe la lista, su cabeza y la posicion de la lista a eliminar
def erase(List, head, pos):
    #Verificamos que la lista no este vacia y la posicion no se ni mayor ni menor a la longitud de la lista
    if head is not None and pos <= List.size and pos > 0:
        #Crearemos nuestra variable contador empezando desde uno la cual nos ayudara a identificar en que posicion de a lista estamos hasta identificar que esta el valor de posicion
        cont = 1
        #Ahora crearemos un nodo curr el cual empezara desde el primer nodo de la lista, ya que este siempre debe de corresponder con la posicion que tenga el contador. Es decir, este terminara ubicandose en la posicion del nodo objetivo
        curr = head.next
        #Por otro lado, tambien necesitamos siempre una referencia al nodo anterior de curr para poder realizar la operacion de eliminado. Por eso, debemos de crear otro nodo llamado a prev el cual siempre sera el anterior a curr
        prev = head
        #Ahora iteraremos hasta el valor de la posicion
        for i in range(1,pos+1):
            #Debemos de preguntar si el contador es igual a la posicion y asi sabremos si ya estamos parados sobre el nodo objetivo o no
            if cont == pos:
                #Si se entra aca, significa que curr ya esta en el nodo objetivo asi que se debe de seguir con la operacion de eliminar
                #Antes de eliminar el nodo objetivo, debemos de asegurarnos siga con el flujo normal. Para esto, haremos que prev apunte al siguiente de curr. 
                #De esta forma , por ningun lado de la lista se tendra acceso ahora a curr, el cual es el nodo objetivo
                prev.next = curr.next
                #Ahora desvincularemos a curr de la lista haciendolo apuntar a nulo
                curr.next = None
                #Decrementamos la longitud de la lista
                List.size -= 1
                #Eliminamos de memoria a curr, es decir, al nodo objetivo
                del curr
            else: 
                #Si se entra aca significa que el contador aun no ha igualado el valor de la posicion recibida.
                #Es decir, curr aun no se ha ubicado en el nodo objetivo
                #Por lo tanto debemos de incrementar el contador y actualizar los nodos auxiliares a sus siguientes
                cont += 1
                prev = curr
                curr = curr.next

#--------------------------------------------------------------